---
output:
  md_document:
    variant: markdown_github
html_preview: false
---


```{r echo=FALSE, results="hide", message=FALSE}
#library("txtplot")
library("badger")
library("ypages")
library("ggplot2")
```


# ChIPseeker: ChIP peak Annotation, Comparison, and Visualization

`r badge_bioc_release("ChIPseeker", "green")`
`r badge_devel("guangchuangyu/ChIPseeker", "green")`
[![Bioc](http://www.bioconductor.org/shields/years-in-bioc/ChIPseeker.svg)](https://www.bioconductor.org/packages/devel/bioc/html/ChIPseeker.html#since)
`r badge_bioc_download("ChIPseeker", "total", "blue")`
`r badge_bioc_download("ChIPseeker", "month", "blue")`

[![Project Status: Active - The project has reached a stable, usable state and is being actively developed.](http://www.repostatus.org/badges/latest/active.svg)](http://www.repostatus.org/#active)
[![codecov](https://codecov.io/gh/GuangchuangYu/ChIPseeker/branch/master/graph/badge.svg)](https://codecov.io/gh/GuangchuangYu/ChIPseeker/)
[![Last-changedate](https://img.shields.io/badge/last%20change-`r gsub('-', '--', Sys.Date())`-green.svg)](https://github.com/GuangchuangYu/ChIPseeker/commits/master)
[![commit](http://www.bioconductor.org/shields/commits/bioc/ChIPseeker.svg)](https://www.bioconductor.org/packages/devel/bioc/html/ChIPseeker.html#svn_source)
[![GitHub forks](https://img.shields.io/github/forks/GuangchuangYu/ChIPseeker.svg)](https://github.com/GuangchuangYu/ChIPseeker/network)
[![GitHub stars](https://img.shields.io/github/stars/GuangchuangYu/ChIPseeker.svg)](https://github.com/GuangchuangYu/ChIPseeker/stargazers)


[![platform](http://www.bioconductor.org/shields/availability/devel/ChIPseeker.svg)](https://www.bioconductor.org/packages/devel/bioc/html/ChIPseeker.html#archives)
[![Build Status](http://www.bioconductor.org/shields/build/devel/bioc/ChIPseeker.svg)](https://bioconductor.org/checkResults/devel/bioc-LATEST/ChIPseeker/)
[![Linux/Mac Travis Build Status](https://img.shields.io/travis/GuangchuangYu/ChIPseeker/master.svg?label=Mac%20OSX%20%26%20Linux)](https://travis-ci.org/GuangchuangYu/ChIPseeker)
[![AppVeyor Build Status](https://img.shields.io/appveyor/ci/Guangchuangyu/ChIPseeker/master.svg?label=Windows)](https://ci.appveyor.com/project/GuangchuangYu/ChIPseeker)



This package implements functions to retrieve the nearest genes around the peak, annotate genomic region of the peak, statstical methods for estimate the significance of overlap among ChIP peak data sets, and incorporate GEO database for user to compare their own dataset with those deposited in database. The comparison can be used to infer cooperative regulation and thus can be used to generate hypotheses. Several visualization functions are implemented to summarize the coverage of the peak experiment, average profile and heatmap of peaks binding to TSS regions, genomic annotation, distance to TSS, and overlap of peaks or genes.


For details, please visit our project website, <https://guangchuangyu.github.io/ChIPseeker>.

+ [Documentation](https://guangchuangyu.github.io/ChIPseeker/documentation/)
+ [Featured Articles](https://guangchuangyu.github.io/ChIPseeker/featuredArticles/)
+ [Feedback](https://guangchuangyu.github.io/ChIPseeker/#feedback)


[![Twitter](https://img.shields.io/twitter/url/https/github.com/GuangchuangYu/ChIPseeker.svg?style=social)](https://twitter.com/intent/tweet?hashtags=ChIPseeker&url=http://bioinformatics.oxfordjournals.org/content/31/14/2382&screen_name=guangchuangyu)



------------------------------------------------------------------------

Please cite the following article when using `ChIPseeker`:

__*Yu G*__, Wang LG and He QY^\*^. ChIPseeker: an R/Bioconductor package for ChIP peak annotation, comparison and visualization. ***Bioinformatics*** 2015, 31(14):2382-2383.

`r badge_doi("10.1093/bioinformatics/btv145", "green")`
`r badge_altmetric("3781087", "green")`

------------------------------------------------------------------------


### Citation

`r badge_citation("9pM33mqn1YgC", "12053363057899219488", "green")`

`r badge_custom("ESI", "Highly Cited Paper", "green")`



```{r echo=F, comment=NA}
d <- get_article_cite_history('DO5oG40AAAAJ', '9pM33mqn1YgC')
## txtplot(d$year, d$cites)
## library("ggalt")
## p <- ggplot(d, aes(year, cites)) + geom_lollipop(point.size=3, point.colour='firebrick') +
p <- ggplot(d, aes(year, cites)) + geom_segment(aes(xend=year, yend=0)) + geom_point(size=3, color='firebrick') +
    theme_minimal() + xlab(NULL) + ylab(NULL) + labs(title="Citation", caption="data from Google Scholar")
ggsave(p, file="docs/citation.png", width=8, height=4)
```


<img src="docs/citation.png" width="890"/>


### Download stats

`r badge_download_bioc("ChIPseeker")`
`r badge_bioc_download("ChIPseeker", "total", "blue")`
`r badge_bioc_download("ChIPseeker", "month", "blue")`



```{r echo=F, message=F, comment=NA}
x <- dlstats::bioc_stats("ChIPseeker")
x <- x[-nrow(x), ] # last month data is not complete
## txtplot(ggtree::Date2decimal(x$end), x$Nb_of_downloads, width=100, height=25)
p <- ggplot(x, aes(end, Nb_of_distinct_IPs)) + geom_point() + geom_line() +
    theme_minimal() +xlab(NULL) + ylab(NULL) +
    labs(title="ChIPseeker", subtitle="Monthly download stats (distinct IPs)", caption="data from Bioconductor")
ggsave(p, file="docs/dlstats.png", width=8, height=4)
```

<img src="docs/dlstats.png" width="890"/>



